csharp/optimized/Program.cs

```diff
@@ -0,0 +1,42 @@
+// Original version by John Taylor
+
+using System;
+using System.IO;
+using System.Collections.Generic;
+using System.Linq;
+
+class Program
+{
+    public sealed class Ref<T> {
```

Comment:
TIL, thanks!

Reply:
What is the purpose of having wrapped the integer anyway?

Reply:
@calledude it's a reference type, so we don't copy a new int every time we get it out of the dictionary - but instead increment the existing one. If you try it without you will see all counts are just 1.

Reply:
What about a `ref struct` then? This would force it to still live on the stack and we would have a pointer to it, perhaps that would make it even more performant? I assume there's a tradeoff in the current solution where having it on the stack does not outweigh the cost of copying the value all the time?

Reply:
To be honest, I haven't looked into that part too much.
A trace I did earlier shows that the actual lookup takes the most time, followed by string creation. (not sure, if the latest code version, but should be close)
![image](https://user-images.githubusercontent.com/8111868/111501654-74fa8380-8745-11eb-9c4d-a53702b0164c.png)

Reply:
> FWIW; According to BenchmarkDotNet `int` is still faster.

@calledude, I'm not sure what exactly you are measuring there, but if you simply removed ref and still use the rest of the optimized code as-is (that is, using `TryGetValue`), your faster code does not work correctly. As @erikbozic already pointed out - all your counters will never increment and always stay at 1.

Yes, extra object indirection adds overhead, but this code is faster than code in "simple" without it exactly because code without it will have to do hash lookup twice: first, to read value and then to store it back again. That is because current Dictionary API does not support atomic increment or in-place value modification.

Reply:
As for you question on `ref struct` - the whole Dictionary lives on the heap and anything you store there has to be boxed and put on a heap, which `ref struct` cannot be.

Reply:
Just double checked and I could've sworn I had made sure I was actually incrementing the value correctly, but I guess not then. 

Not my proudest moment... :) Sorry about that!